cygwin/Platform.c

```diff
@@ -0,0 +1,337 @@
+   { .name = "12 SIGSYS",    .number = 12 },
+   { .name = "13 SIGPIPE",   .number = 13 },
+   { .name = "14 SIGALRM",   .number = 14 },
+   { .name = "15 SIGTERM",   .number = 15 },
+   { .name = "16 SIGURG",    .number = 16 },
+   { .name = "17 SIGSTOP",   .number = 17 },
+   { .name = "18 SIGTSTP",   .number = 18 },
+   { .name = "19 SIGCONT",   .number = 19 },
+   { .name = "20 SIGCHLD",   .number = 20 },
+   { .name = "21 SIGTTIN",   .number = 21 },
+   { .name = "22 SIGTTOU",   .number = 22 },
+   { .name = "23 SIGIO",     .number = 23 },
+   { .name = "24 SIGXCPU",   .number = 24 },
+   { .name = "25 SIGXFSZ",   .number = 25 },
+   { .name = "26 SIGVTALRM", .number = 26 },
+   { .name = "27 SIGPROF",   .number = 27 },
+   { .name = "28 SIGWINCH",  .number = 28 },
+   { .name = "29 SIGINFO",   .number = 29 },
+   { .name = "30 SIGUSR1",   .number = 30 },
+   { .name = "31 SIGUSR2",   .number = 31 },
+};
+
+const unsigned int Platform_numberOfSignals = ARRAYSIZE(Platform_signals);
+
+const MeterClass* const Platform_meterTypes[] = {
+   &CPUMeter_class,
+   &ClockMeter_class,
+   &DateMeter_class,
+   &DateTimeMeter_class,
+   &LoadAverageMeter_class,
+   &LoadMeter_class,
+   &MemoryMeter_class,
+   &SwapMeter_class,
+   &MemorySwapMeter_class,
+   &TasksMeter_class,
+   &UptimeMeter_class,
+   &BatteryMeter_class,
+   &HostnameMeter_class,
+   &SysArchMeter_class,
+   &AllCPUsMeter_class,
+   &AllCPUs2Meter_class,
+   &AllCPUs4Meter_class,
+   &AllCPUs8Meter_class,
+   &LeftCPUsMeter_class,
+   &RightCPUsMeter_class,
+   &LeftCPUs2Meter_class,
+   &RightCPUs2Meter_class,
+   &LeftCPUs4Meter_class,
+   &RightCPUs4Meter_class,
+   &LeftCPUs8Meter_class,
+   &RightCPUs8Meter_class,
+   &BlankMeter_class,
+   NULL
+};
+
+bool Platform_init(void) {
+   /* no platform-specific setup needed */
+   return true;
+}
+
+void Platform_done(void) {
+   /* no platform-specific cleanup needed */
+}
+
+void Platform_setBindings(Htop_Action* keys) {
+   /* no platform-specific key bindings */
+   (void) keys;
+}
+
+int Platform_getUptime(void) {
+   double uptime = 0;
+   FILE* fd = fopen(PROCDIR "/uptime", "r");
+   if (fd) {
+      int n = fscanf(fd, "%64lf", &uptime);
+      fclose(fd);
+      if (n <= 0) {
+         return 0;
+      }
+   }
+   return floor(uptime);
+}
+
+void Platform_getLoadAverage(double* one, double* five, double* fifteen) {
+   FILE* fd = fopen(PROCDIR "/loadavg", "r");
+   if (!fd)
+      goto err;
+
+   double scanOne, scanFive, scanFifteen;
+   int r = fscanf(fd, "%lf %lf %lf", &scanOne, &scanFive, &scanFifteen);
+   fclose(fd);
+   if (r != 3)
+      goto err;
+
+   *one = scanOne;
+   *five = scanFive;
+   *fifteen = scanFifteen;
+   return;
+
+err:
+   *one = NAN;
+   *five = NAN;
+   *fifteen = NAN;
+}
+
+pid_t Platform_getMaxPid(void) {
+   // https://github.com/cygwin/cygwin/blob/a9e8e3d1cb8235f513f4d8434509acf287494fcf/winsup/cygwin/local_includes/pinfo.h#L223
+   return 65536;
+}
```

Comment:
Note the limitations regarding Cygwin here. On native Windows the story is [a bit different](https://learn.microsoft.com/en-us/answers/questions/70930/maximum-value-of-process-id) ([Reymond Chen](https://devblogs.microsoft.com/oldnewthing/20040223-00/?p=40503)) …

Reply:
> Note the limitations regarding Cygwin here. On native Windows the story is a bit different

but, this is a cygwin port, not native windows port...